pkg/apicheck/check.go

```diff
@@ -167,15 +180,43 @@ func (c *ApiConnectivityCheck) handleError() bool {
 			if apiErrorsResponsesSum > nrAllNodes/2 { //already reached more than 50% of the nodes and all of them returned api error
 				//assuming this is a control plane failure as others can't access api-server as well
 				c.config.Log.Info("More than 50% of the nodes couldn't access the api-server, assuming this is a control plane failure")
-				return true
+				return peers.Response{IsHealthy: true, Reason: peers.HealthyBecauseMostPeersCantAccessAPIServer}
 			}
 		}
 
 	}
 
 	//we asked all peers
 	c.config.Log.Error(fmt.Errorf("failed health check"), "Failed to get health status peers. Assuming unhealthy")
-	return false
+	return peers.Response{IsHealthy: false, Reason: peers.UnHealthyBecauseNodeIsIsolated}
+}
+
+func (c *ApiConnectivityCheck) canOtherMastersBeReached() bool {
+	nodesToAsk := c.config.Peers.GetPeersAddresses(peers.Master)
+	numOfMasterPeers := len(nodesToAsk)
+	if nodesToAsk == nil || numOfMasterPeers == 0 {
+		c.config.Log.Info("Peers list is empty and / or couldn't be retrieved from server, other masters can't be reached")
+		return false
+	}
+	chosenNodesAddresses := c.popNodes(&nodesToAsk, numOfMasterPeers)
+
+	nrAddresses := len(chosenNodesAddresses)
+	responsesChan := make(chan selfNodeRemediation.HealthCheckResponseCode, nrAddresses)
+	var wg sync.WaitGroup
+	wg.Add(len(chosenNodesAddresses))
+	for _, address := range chosenNodesAddresses {
+		tmpAddress := address
+		go func() {
+			defer wg.Done()
+			c.getHealthStatusFromPeer(tmpAddress, responsesChan)
+		}()
+
+	}
+	wg.Wait()
+	//We are not expecting API Server connectivity at this stage, however an API Error is an indication to communication with the peer (peer is communicating with current node that it was unable to reach the API server)
+	_, _, apiErrorsResponses, _ := c.sumPeersResponses(nrAddresses, responsesChan)
```

Comment:
ping @mshitrit

Reply:
>Why don't we expect any API server connectivity at thi stage?

This is the branch where we are asking the peers whether we are healthy or not, we get here because there is no  API server connectivity (otherwise there was no need to ask the peers).

>Would it maybe be to better check for noResponse count (last return value)?

Basically this  is the same way we used to do it for worker nodes, so I'm not eager to change it without extensive testing.
IIUC `noResponse` maps in this context to failedRequest (due to issues with certs, failing to initilize the client etc...) which in a nutshell means that the peer wasn't pinged.

Reply:
> we get here because there is no API server connectivity

The **current** node doesn't have connectivity. Peers can still have it, not?

> Basically this is the same way we used to do it for worker nodes

I don't think so. The main interest in getWorkerPeersResponse is to know if peers think we are healthy or not.
Here the interest is if we can reach peers at all, at least according to the function name, and according to code not looking at healthy or unhealthy results. Which makes sense.

My point is: IMHO it is possible that peers can reach the api server, and then apiErrorsResponses will be 0. Then we return "peers could not be reached" even though they could be reached. That's why I think noResponse is the better value to look at.

> IIUC noResponse maps in this context to failedRequest (due to issues with certs, failing to initilize the client etc...) which in a nutshell means that the peer wasn't pinged.

Correct, there can be other reasons for failedRequest than a failed "ping". We can add another response code for those cases, e.g. RequestPreparationFailed

Reply:
>The current node doesn't have connectivity. Peers can still have it, not?

That is true but this is not the intention.
Checking [IsControlPlaneHealthy](https://github.com/medik8s/self-node-remediation/pull/48/files#diff-7e10780971a8c80e5b146a094d3adb33d0e16a193ad19319e6bdcba30b43732eR60) method, we will consider the result of this method only in case that workers peers couldn't be reached and we think the node might be isolated, otherwise we will remediate according to worker peers result making the result of this method non relevant.
I guess the comment is indeed misleading - I'll update it.

Reply:
Sorry, I can't follow.
When the control plane peers have API server connectivity, apiErrorsResponses will be 0. With the current implementation that is understood as "this node is isolated" from its control plane peers. But it isn't, it successfully connected to the control plane peers. 🤔

Reply:
or to rephrase:
apiErrorsResponses > 1 is an indication for connectivity, yes
but apiErrorsResponses == 0 isn't an indication for not having connectivity